File: backend/src/schemas/agent_knowledge_bindings.py
"""Knowledge dataset bindings stored on agent.config_json.knowledge_bindings."""

from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, Field


class AgentKnowledgeBindings(BaseModel):
    dataset_ids: list[UUID] = Field(default_factory=list)

    def is_empty(self) -> bool:
        return not self.dataset_ids


def parse_knowledge_bindings(config_json: dict | None) -> AgentKnowledgeBindings:
    raw = (config_json or {}).get("knowledge_bindings") or {}
    if not isinstance(raw, dict):
        return AgentKnowledgeBindings()
    dataset_ids = [UUID(str(x)) for x in raw.get("dataset_ids") or []]
    return AgentKnowledgeBindings(dataset_ids=dataset_ids)


def merge_knowledge_bindings_into_config(
    config_json: dict | None,
    bindings: AgentKnowledgeBindings | None,
) -> dict:
    base = dict(config_json or {})
    if bindings is not None:
        base["knowledge_bindings"] = {
            "dataset_ids": [str(x) for x in bindings.dataset_ids],
        }
    return base

File: frontend/src/lib/agent-knowledge-bindings.ts
import type { AgentKnowledgeBindings } from "@/types";

export const EMPTY_KNOWLEDGE_BINDINGS: AgentKnowledgeBindings = { dataset_ids: [] };

export function parseKnowledgeBindings(
  config?: Record<string, unknown> | null
): AgentKnowledgeBindings {
  const raw = config?.knowledge_bindings;
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    return { dataset_ids: [] };
  }
  const o = raw as Record<string, unknown>;
  return {
    dataset_ids: Array.isArray(o.dataset_ids) ? (o.dataset_ids as string[]) : [],
  };
}

export function mergeKnowledgeBindingsIntoConfig(
  config: Record<string, unknown> | null | undefined,
  bindings: AgentKnowledgeBindings
): Record<string, unknown> {
  return {
    ...(config ?? {}),
    knowledge_bindings: {
      dataset_ids: [...bindings.dataset_ids],
    },
  };
}
